File: 백준/Gold/7579. 앱/앱.cc
#include <iostream>
#include <algorithm>

using namespace std;

struct app {
	int m, cost;

	bool operator< (app other) {
		return cost < other.cost;
	}
};

int main() {
	int n, m;
	int temp[100];
	app apps[100];

	cin >> n >> m;

	for (int i = 0; i < n; i++) {
		cin >> temp[i];
	}

	for (int i = 0; i < n; i++) {
		int num;
		cin >> num;
		apps[i] = { temp[i], num };
	}

	sort(apps, apps + n);

	int dp[10001][2];
	int state = 0;

	for (int i = 0; i <= n * 100; i++) {
		dp[i][1] = 0;
	}

	for (int i = 0; i < n; i++) {
		int other = (state + 1) % 2;

		for (int j = 0; j <= n * 100; j++) {
			if (j<apps[i].cost || dp[j][other]>=dp[j - apps[i].cost][other]+apps[i].m) {
				dp[j][state] = dp[j][other];
			}
			else {
				dp[j][state] = dp[j - apps[i].cost][other] + apps[i].m;
			}

			if (dp[j][state] >= m) {
				break;
			}
		}
		state = (state + 1) % 2;
	}
	state = (state + 1) % 2;

	for (int i = 0; i <= n * 100; i++) {
		if (dp[i][state] >= m) {
			cout << i << "\n";
			break;
		}
	}

	return 0;
}

File: 백준/Gold/7579. 앱/앱.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main {

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		PriorityQueue<App> apps = new PriorityQueue<>((e1, e2) -> e1.cost-e2.cost);

		int temp[] = new int[N];
		
		st = new StringTokenizer(br.readLine());
		for(int i=0; i<N; i++) {
			temp[i] = Integer.parseInt(st.nextToken());
		}
		
		st = new StringTokenizer(br.readLine());
		for(int i=0; i<N; i++) {
			apps.offer(new App(temp[i], Integer.parseInt(st.nextToken())));
		}
		
		int dp[][] = new int[10001][2];
		int state = 0;

		for (int i = 0; i <= N * 100; i++) {
			dp[i][1] = 0;
		}

		for (int i = 0; i < N; i++) {
			int other = (state + 1) % 2;
			App app = apps.poll();

			for (int j = 0; j <= N * 100; j++) {
				
				if (j<app.cost || dp[j][other]>=dp[j - app.cost][other]+app.m) {
					dp[j][state] = dp[j][other];
				}
				else {
					dp[j][state] = dp[j - app.cost][other] + app.m;
				}

				if (dp[j][state] >= M) {
					break;
				}
			}
			state = (state + 1) % 2;
		}
		state = (state + 1) % 2;
		
		for (int i = 0; i <= N * 100; i++) {
			if (dp[i][state] >= M) {
				System.out.println(i);
				break;
			}
		}
	}

	static class App {
		int m, cost;
		public App(int m, int cost) {
			this.m = m;
			this.cost = cost;
		}
	}
}
